### PR TITLE
Connect database

### PR DESCRIPTION
**Pull request by katiasuya**

### CODE REVIEW

**Comment by katiasuya** [comment]
Change application layout

File: internal/server/handlers.go [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+	hashedPwd, err := s.repo.GetUserPassword(lr.Username)
+	if err != nil {
+		RespondErr(w, http.StatusUnauthorized, err)
+		return
+	}
+
+	if !hash.CheckPasswordHash(lr.Password, hashedPwd) {
+		RespondErr(w, http.StatusUnauthorized, errors.New("wrong password"))
+		return
+	}
```
Comment (papahmsolo) [inline internal/server/handlers.go:101]: In all of these cases let's return 400 with `invalid username or password`
https://stackoverflow.com/questions/14922130/which-error-message-is-better-when-users-entered-a-wrong-password

File: internal/server/handlers.go [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+	if err := ValidateRequest(name, sourceFormat, targetFormat); err != nil {
+		RespondErr(w, http.StatusBadRequest, err)
+		return
+	}
+
+	requestID, err := s.repo.MakeRequest(name, sourceFormat, targetFormat, "some location", "992dee5c-b4e3-49f8-9d4c-8903fa2284c9")
+	if err != nil {
+		RespondErr(w, http.StatusBadRequest, err)
+		return
+	}
```
Comment (papahmsolo) [inline internal/server/handlers.go:132]: Shouldn't it be 500 in this case?

File: internal/repository/postgres.go [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	err = db.Ping()
+	if err != nil {
+		return nil, err
+	}
+
+	return db, nil
```
Comment (papahmsolo) [inline internal/repository/postgres.go:24]: You can simply `return db, err`

File: internal/repository/repository.go [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+}
+
+// GetUserPassword retrieves the database hashed password of a user.
+func (r *Repository) GetUserPassword(username string) (string, error) {
+	var password string
+	const getPasswordByUsername = `SELECT password FROM converter."user" WHERE username=$1;`
+	err := r.db.QueryRow(getPasswordByUsername, username).Scan(&password)
+	if err == sql.ErrNoRows {
+		return "", errors.New("the user with the given username does not exist")
+	}
```
Comment (papahmsolo) [inline internal/repository/repository.go:45]: You'll need to create an error type for this error to distinguish between "no such user" and failed DB query in the handler

**Review by papahmsolo — CHANGES_REQUESTED** [review]
Looks very good!

One note: in the handlers you can name your nested types as `request` and `response`, thus you will have the same name across all handlers.

File: internal/server/handlers.go [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+
+	userID, err := s.repo.InsertUser(sr.Username, sr.Password)
+	if err != nil {
+		RespondErr(w, http.StatusInternalServerError, err)
+		return
+	}
```
[the remark this answers is not in the document]
Comment (katiasuya) [reply internal/server/handlers.go:65]: I've handled this situation in InsertUser function as pq.Error

File: internal/repository/postgres.go [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+func NewPostgresDB(c *config.Config) (*sql.DB, error) {
+	pqInfo := fmt.Sprintf("host=%s port=%d user=%s password=%s dbname=%s sslmode=%s",
+		c.Host, c.Port, c.Username, c.Password, c.DBName, c.SSLMode)
+	db, err := sql.Open("postgres", pqInfo)
+	if err != nil {
+		return nil, err
+	}
+
+	err = db.Ping()
+	return db, err
```
Comment (papahmsolo) [inline internal/repository/postgres.go:20]: `return db, db.Ping()`

Comment (papahmsolo) [reply internal/repository/postgres.go:20]: But your way is also fine

File: internal/server/handlers.go [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+		RespondErr(w, http.StatusUnauthorized, err)
+		return
+	}
+	defer r.Body.Close()
+
+	hashedPwd, err := s.repo.GetUserPassword(req.Username)
+	if err == repository.ErrNoSuchUser {
+		RespondErr(w, http.StatusNotFound, err)
+		return
+	}
```
Comment (papahmsolo) [inline internal/server/handlers.go:112]: We shouldn't return anything but 401 in such cases 
https://stackoverflow.com/questions/14922130/which-error-message-is-better-when-users-entered-a-wrong-password

File: internal/server/handlers.go [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+		RespondErr(w, http.StatusInternalServerError, err)
+		return
+	}
+	defer out.Close()
+
+	_, err = io.Copy(out, resp.Body)
+	if err != nil {
+		RespondErr(w, http.StatusInternalServerError, err)
+		return
+	}
```
Comment (papahmsolo) [inline internal/server/handlers.go:211]: Not sure, why would download create files :)
Don't worry about it for now, we'll implement file storage in the next step

File: internal/repository/repository.go [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+	var hrs []HistoryResponse
+	for rows.Next() {
+		err = rows.Scan(&hr.ID, &hr.AudioName, &hr.SourceFormat, &hr.TargetFormat, &hr.Created, &hr.Updated, &hr.Status)
+		if err != nil {
+			return nil, err
+		}
+		hrs = append(hrs, hr)
+	}
+	err = rows.Err()
+	return hrs, err
```
Comment (papahmsolo) [inline internal/repository/repository.go:111]: same here: `return hrs, rows.Err()`

**Review by papahmsolo — CHANGES_REQUESTED** [review]
Looks good, let's fix some things

**Review by papahmsolo — APPROVED** [review]

**Pull request merged**